Feature/gptl

christopherwharrop-noaa opened the pull request:
## Proposed changes

Add support for profiling with [GPTL](https://github.com/jmrosinski/GPTL).  Closes #13.

This pull request adds two build options for profiling with GPTL.

The first option, ENABLE_GPTL, turns on profiling with GPTL and is enabled with the ENABLE_GPTL cmake flag:
```
cmake -DENABLE_GPTL=1 ......
```
The second option, ENABLE_AUTOPROFILING, turns on GPTL's autoprofiling feature if ENABLE_GPTL is also enabled:
```
cmake -DENABLE_GPTL=1 -DENABLE_AUTOPROFILING=1 ....
```

The CI MacOS and Ubuntu build tests have been updated to install GPTL and add a build test with `-DENABLE_GPTL=1 -DENABLE_AUTOPROFILING=1`

## Types of changes

What types of changes does your code introduce?

- [ ] Bugfix (Change which fixes an issue)
- [x] New feature (Change which adds functionality)
- [ ] Enhancement to existing functionality (non-breaking change which improves existing code or documentation)
- [ ] Breaking change (fix or feature that changes results and requires updates to test baselines)
- [ ] Documentation Update

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/NOAA-GSL/SENA-c_sw/blob/develop/CONTRIBUTING.md) document
- [x] I have ensured all my changes contribute to a common purpose
- [x] I have verified my change does not add debug code
- [x] I have verified my change does not add code that is commented out
- [x] I have verified unit tests pass locally with my changes
- [x] I have verified my changes adhere to style guidelines
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

For now, this is a draft PR.  Once documentation is added, and CI testing is verified to work properly, it will be ready for review.

christopherwharrop-noaa committed:
Add basic support for GPTL timing

Add build options for GPTL and AUTOPROFILING.
Both options are OFF by default.

If GPTL is 1 or "on", it will build with support for GPTL timers.
If AUTOPROFILING is 1 or "on", it will build with autoprofiling enabled.

sw_driver.f90 had to be renamed with .F90 so that preprocessing
for "#ifdef GPTL_ENABLED" would be done automatically by the compiler.

christopherwharrop-noaa committed:
Add CI test for GPTL support

christopherwharrop-noaa committed:
Add basic check for GPTL output

christopherwharrop-noaa committed:
Add GPTL instructions to README

christopherwharrop-noaa committed:
Remove sw_driver.f90 that was renamed to sw_driver.F90

christopherwharrop-noaa committed:
Remove -rdynamic autoprofiling flag

christopherwharrop-noaa committed:
Fix typos in README

christopherwharrop-noaa committed:
Add GPTL install and test for MacOS build

LyndStringer commented on ref/README.md:
```diff
@@ -79,6 +84,27 @@ On many systems, the above will suffice. However, some systems will require you
+If GPTL is installed and loaded onto system paths, this kernel may be built with support for
+GPTL profiling. There are two options.
+
+The first option enables GPTL and allows you to	insert your own	custom GPTL timers into the code
```
I would suggest writing about autoprofiling first, because that is the most common use.  The "insert your own custom GLTP timers" is rarer.

LyndStringer commented on ref/README.md:
```diff
@@ -79,6 +84,27 @@ On many systems, the above will suffice. However, some systems will require you
 
+### Building with GPTL profiling support
+
+If GPTL is installed and loaded onto system paths, this kernel may be built with support for
```
I suggest inserting a sentence here explaining what GPTL is and does.  As written, the implicit expectation is that the reader already knows what GPTL is and does.

LyndStringer commented on ref/README.md:
````diff
@@ -79,6 +84,27 @@ On many systems, the above will suffice. However, some systems will require you
+$ cmake -DCMAKE_BUILD_TYPE=debug -DENABLE_GPTL=1 ..
+```
+
+The second option enables GPTL autoprofiling, which will give you timings for all subroutine calls
````
With autoprofiling, where is the output file? format?

LyndStringer commented on ref/README.md:
```diff
@@ -79,6 +84,27 @@ On many systems, the above will suffice. However, some systems will require you
 find dependencies, particularly if software depencencies are not installed in standard locations.
 See below for more information.
 
+### Building with GPTL profiling support
```
I would suggest adding a link to the GPTL documentation for the hardy souls that want to insert their own GPTL calls.

LyndStringer reviewed (commented):
Lots of cmake code.

christopherwharrop-noaa committed:
Modify README for clarity

christopherwharrop-noaa committed:
Fix link to GPTL optional dependency in README

robgpita reviewed (dismissed):
Looks good to me! Couldn't find anything that would warrant a change.

middlecoff reviewed (commented):
I don't see any changes to sw_core.F90. Am I missing them? Previously sw_core.F90 had commented out calls to GPTL, are those to be left as comments?

christopherwharrop-noaa commented:
@middlecoff - I didn't realize there were commented calls to start/stop GPTL timers in `sw_core.F90`.  My assumption was that users would use the autoprofiling to start and then add, if needed, their own custom timers to meet the requirements of their particular goal.  So, no timers were added to the code because I was expecting the users to do that themselves.

We could add some basic timers, perhaps the ones that are already there but commented out, as a starting point.  How do others feel about that? (@LyndStringer , @robgonzalezpita , @markgovett )

robgpita commented:
@christopherwharrop-noaa If we did go this route, `gptl` would no longer be an "optional dependency".  It would be nice to see the timing information from `gptl` for the `c_sw` subroutine; however, this is already accomplished with the timing information being printed from the driver code.  I'll defer to the group in whether we want to make `gptl` a required dependency.

christopherwharrop-noaa commented:
@robgonzalezpita - All GPTL calls will need to enclosed in

```
#ifdef ENABLE_GPTL
  if (do_profile == 1) then
    ! Start/stop the timer here
  end if
#endif
```

All GPTL code that becomes part of the repository must be enclosed in that.  The `#ifdef` makes sure the code is only compiled if GPTL support was enabled at build time.  the `if (do_profile==1)` makes sure that the code is only called if the runtime switch is turned on.  Note that I haven't pushed the code to do that second part yet, but I have it written and tested.

All in all, it will still be an optional dependency if we do it correctly.

I think the advantage of GPTL is that it is specifically designed for profiling, so while we do have the existing use of the clock to time the kernel, GPTL is better positioned to do this in a more general way with attention to all the nuances of profiling.

christopherwharrop-noaa committed:
Add capability to turn profiling on/off at runtime

Add namelist option for turning profiling on/off at runtime.
Wrap calls to GPTL in check for runtime option.
Add optional argument to build.sh convenience script for GPTL support.

christopherwharrop-noaa committed:
Update README with GPTL instructions for build.sh

christopherwharrop-noaa committed:
Fix CI test for GPTL support so profiling is turned on at runtime

christopherwharrop-noaa committed:
Add diagnostic info to CI GPTL test

christopherwharrop-noaa committed:
Fix typo in CI tests for GPTL support

christopherwharrop-noaa committed:
Fix sed command to GPTL CI test will work on MacOS build

robgpita commented:
@christopherwharrop-noaa Good to know you've already written and tested the code so that we can keep GPTL as an optional dependency.  If thats the case, I don't see why we shouldn't add it, barring any major drawbacks. The only ones I can think of would be added complexity, and certain use cases not needing the profiling?

LyndStringer reviewed (dismissed):
Looks good.

middlecoff reviewed (commented):
I don't see sw_core.F90 in the changed code (am I missing it?), are you leaving the commented GPTL calls in sw_core.F90?

christopherwharrop-noaa commented:
@middlecoff - I responded to that query above but some intervening discussion happened so perhaps it got lost.  

You are not missing anything.  I intentionally did not include adding timers (I didn't know there were commented out calls in the code) because I had assumed it would be up to the user to insert them for their own purposes.  

However, after seeing those calls that you pointed out, I've gone ahead and included them in another revision that I haven't committed/pushed yet. I was waiting to see how folks felt about the pros/cons of having those basic calls in.  I think I like having them in there for two reasons:

1. It means that the basic parts of the code will be timed by default.
2. It provides users with an example of how to add their own timers.

So, I think I'm going to add that.  Stand by...

christopherwharrop-noaa committed:
Add simple GPTL timers for each code block

christopherwharrop-noaa commented:
I've just added basic timers for each major subroutine.  When using GPTL with manual profiling (as opposed to autoprofiling) the user will get output that looks like this:

```
Stats for thread 0:
                          Called  Recurse     Wall      max      min   selfOH parentOH
  kernel                       1     -    5.90e-03 5.90e-03 5.90e-03    0.000    0.000
    c_sw                      32     -    5.65e-03 3.78e-04 1.66e-04    0.000    0.000
      d2a2c_vect              32     -    1.46e-03 1.32e-04 4.10e-05    0.000    0.000
      divergence_corner       32     -    7.02e-04 4.10e-05 2.10e-05    0.000    0.000
      fill2_4corners          64     -    2.50e-05 1.20e-05 0.00e+00    0.000    0.000
      fill_4corners           64     -    1.40e-05 3.00e-06 0.00e+00    0.000    0.000
Overhead sum =  1.76e-05 wallclock seconds
Total calls  = 225
```

When using autoprofiling, the user will still get info for those same timers, but will get everything else, too:

```
Stats for thread 0:
                                                  Called  Recurse     Wall      max      min   selfOH parentOH
  __sw_core_mod_MOD_read_state                         1     -       0.025    0.025    0.025    0.000    0.000
*   __netcdfmodule_MOD_open_file                       2     -    7.09e-03 4.51e-03 2.58e-03    0.000    0.000
    __netcdfmodule_MOD_read_global_int                12     -    4.90e-05 4.00e-05 0.00e+00    0.000    0.000
    __netcdfmodule_MOD_read_global_real                1     -    7.00e-06 7.00e-06 7.00e-06    0.000    0.000
    __netcdfmodule_MOD_read_global_logical             4     -    7.00e-06 3.00e-06 1.00e-06    0.000    0.000
    __netcdfmodule_MOD_read_dimension                  6     -    1.90e-05 1.20e-05 1.00e-06    0.000    0.000
    __sw_core_mod_MOD_allocate_state                   1     -    5.39e-03 5.39e-03 5.39e-03    0.000    0.000
*     __sw_core_mod_MOD_deallocate_state               2     -    8.20e-04 8.17e-04 3.00e-06    0.000    0.000
    __netcdfmodule_MOD_read_2d_real                   19     -    1.90e-04 6.90e-05 6.00e-06    0.000    0.000
*     __netcdfmodule_MOD_nc_check                    205     -    1.40e-05 1.00e-06 0.00e+00    0.000    0.000
    __netcdfmodule_MOD_read_3d_real                   17     -       0.013 3.40e-03 4.20e-05    0.000    0.000
*   __netcdfmodule_MOD_close_file                      2     -    4.77e-03 2.96e-03 1.81e-03    0.000    0.000
  __sw_core_mod_MOD_print_state                        2     -       0.012 5.93e-03 5.78e-03    0.000    0.000
    __sw_core_mod_MOD_print_2d_variable               38     -    4.45e-04 5.20e-05 8.00e-06    0.000    0.000
    __sw_core_mod_MOD_print_3d_variable               34     -       0.011 4.46e-04 3.10e-05    0.000    0.000
  kernel                                               1     -    6.28e-03 6.28e-03 6.28e-03    0.000    0.000
    __sw_core_mod_MOD_c_sw                            31     -    6.08e-03 4.96e-04 1.80e-04    0.000    0.000
      c_sw                                            31     -    5.95e-03 4.02e-04 1.78e-04    0.000    0.000
        __sw_core_mod_MOD_d2a2c_vect                  31     -    1.77e-03 1.47e-04 5.00e-05    0.000    0.000
          d2a2c_vect                                  31     -    1.73e-03 1.30e-04 4.90e-05    0.000    0.000
            __sw_core_mod_MOD_edge_interpolate4     2480     -    8.20e-05 1.00e-06 0.00e+00    0.000    0.000
        __sw_core_mod_MOD_divergence_corner           31     -    7.11e-04 4.20e-05 2.10e-05    0.000    0.000
          divergence_corner                           31     -    6.76e-04 3.60e-05 2.00e-05    0.000    0.000
        __sw_core_mod_MOD_fill2_4corners              62     -    4.70e-05 1.20e-05 0.00e+00    0.000    0.000
          fill2_4corners                              62     -    2.90e-05 9.00e-06 0.00e+00    0.000    0.000
        __sw_core_mod_MOD_fill_4corners               62     -    2.40e-05 1.00e-06 0.00e+00    0.000    0.000
          fill_4corners                               62     -    1.10e-05 1.00e-06 0.00e+00    0.000    0.000
  __sw_core_mod_MOD_write_state                        1     -       0.022    0.022    0.022    0.000    0.000
    __netcdfmodule_MOD_write_global_character          2     -    4.90e-05 4.50e-05 4.00e-06    0.000    0.000
    __netcdfmodule_MOD_write_global_int               12     -    2.20e-05 8.00e-06 1.00e-06    0.000    0.000
    __netcdfmodule_MOD_write_global_real               1     -    5.00e-06 5.00e-06 5.00e-06    0.000    0.000
    __netcdfmodule_MOD_write_global_logical            4     -    9.00e-06 3.00e-06 2.00e-06    0.000    0.000
    __netcdfmodule_MOD_define_dim                      6     -    1.60e-05 8.00e-06 1.00e-06    0.000    0.000
    __netcdfmodule_MOD_define_var_2d_real             19     -    2.09e-04 7.20e-05 5.00e-06    0.000    0.000
    __netcdfmodule_MOD_define_var_3d_real             17     -    1.32e-04 1.00e-05 5.00e-06    0.000    0.000
    __netcdfmodule_MOD_define_off                      1     -    6.62e-03 6.62e-03 6.62e-03    0.000    0.000
    __netcdfmodule_MOD_write_var_2d_real              19     -    4.00e-04 1.62e-04 1.20e-05    0.000    0.000
    __netcdfmodule_MOD_write_var_3d_real              17     -    8.38e-03 6.19e-04 4.00e-05    0.000    0.000
Overhead sum =  0.000252 wallclock seconds
Total calls  = 3360
```

christopherwharrop-noaa commented:
@middlecoff, @robgonzalezpita, @LyndStringer - Sorry, everyone, for so many updates after your approvals.  Many different issues needed to be addressed.  If you could please see the last comment above and re-approve or make a suggestion, that would be appreciated.  I'm hoping this will be the final go around for this.  I greatly appreciated all your comments and suggestions that helped get this into shape.

middlecoff reviewed (approved):
Looks good!

robgpita reviewed (approved):
Looks good. Thanks for all your work on adding the GPTL support!

LyndStringer commented:
Is GPTL on by default, in other words, mandatory to run SENA-c_sw?  If so, I'm concerned we'll have trouble on new architectures.  I'll move these changes to the ARM or AMD machines and see if it will execute.

christopherwharrop-noaa commented:
@LyndStringer - No, it is off by default.  In the top-level cmake, you can see:

```
option( ENABLE_GPTL "Use GPTL profiling library (if available)" OFF )
option( ENABLE_AUTOPROFILING "Enable function-based autoprofiling with GPTL (if available)" OFF )
```

And it won't even look for GPTL unless the user has enabled it for the build because:

```
if ( ENABLE_GPTL )
  find_package( GPTL )
  add_compile_definitions(ENABLE_GPTL)
endif()

if ( GPTL_FOUND AND ENABLE_AUTOPROFILING )
  set(CMAKE_Fortran_FLAGS "${CMAKE_Fortran_FLAGS} ${Fortran_AUTOPROFILING_FLAGS}")
endif()
```

So, the kernel will work just fine on machines without a GPTL installation.  This was made completely optional for the exact concern you expressed.

christopherwharrop-noaa commented:
To further verify that it works without GPTL, there are the CI tests that build and run and test it without a GPTL installed.  As you can see those are passing on both MacOS and Ubuntu.

LyndStringer reviewed (approved):
My concerns about a "mandatory GPTL" have been assuaged [nice SAT word!].

christopherwharrop-noaa commented:
@LyndStringer - We will need to carefully vet any future PRs that include calls to GPTL to make sure they are properly enclosed inside:

```
#ifdef ENABLE_GPTL
  if (do_profile == 1) then
    ! call to GPTL goes here
  end if
#endif
```

Any calls that are not wrapped as above WILL cause the code to fail when not built with GPTL support.  However, since our CI tests now check builds with and without GPTL, such instances should not be able to sneak in.
